File: php/src/Level2.php
<?php declare(strict_types=1);

namespace DJB;

class Level2 extends Level
{
    public function __construct()
    {
        parent::__construct(2);
    }

    public function run()
    {
        $cars = [];
        foreach($this->input->cars as $car) {
            $cars[$car->id] = $car;
        }
        
        $rentals = [];
        foreach($this->input->rentals as $rental) {
            $car = $cars[$rental->car_id];
            $duration = (new \DateTime($rental->end_date))->diff(new \DateTime($rental->start_date))->format("%a") + 1;
            
            $price = $rental->distance * $car->price_per_km + $this->durationPrice($duration, $car->price_per_day);
            $rentals[] = (object)[
                'id' => $rental->id,
                'price' => $price
            ];
        }

        $output = (object)[
            'rentals' => $rentals
        ];

        return $output;       
    }

    private function durationPrice(int $duration, int $price_per_day)
    {
        $steps = [
            1 => 10,
            4 => 30,
            10 => 50
        ];

        krsort($steps);

        $price = 0;
        foreach($steps as $days => $reduction) {
            if($duration <= $days) {
                continue;
            }

            $price += ($duration - $days) * $price_per_day * (100 - $reduction) / 100;
            $duration -= ($duration - $days);

        }
        
        return $price + $duration * $price_per_day;
    }
}

File: elixir/lib/level2.ex
defmodule Drivy.Level2 do
  import Drivy

  def run(input) do
    cars = to_map(input["cars"])
    rentals = input["rentals"]

    rentals =
      Enum.map(rentals, fn r ->
        car = Map.get(cars, r["car_id"])
        %{"id" => r["id"], "price" => get_price(car, r)}
      end)

    %{"rentals" => rentals}
  end

  defp get_price(car, rental) do
    car["price_per_km"] * rental["distance"] +
      get_reduction(to_days(rental["start_date"], rental["end_date"]), car["price_per_day"])
  end

  def get_reduction(duration, price_per_day) do
    cond do
      duration > 10 -> (duration - 10) * price_per_day * 0.5 + get_reduction(10, price_per_day)
      duration > 4 -> (duration - 4) * price_per_day * 0.7 + get_reduction(4, price_per_day)
      duration > 1 -> (duration - 1) * price_per_day * 0.9 + get_reduction(1, price_per_day)
      true -> duration * price_per_day
    end
  end
end
